text/0143-versioning-and-stability.md

```diff
@@ -0,0 +1,130 @@
+
+**Removed -** a deprecated signal is no longer supported, and is removed.
+
+All signal components may become stable together, or one by one in the following order: API, SDK, Semantic Conventions, Contrib.
+
+### Stability
+
+Once a signal component is marked as stable, the following rules apply until the end of that signal’s existence.
+
+**API Stability -**
+No backward-incompatible changes to the API are allowed unless the major version number is incremented. All existing API calls must continue to compile and function against all future minor versions of the same major version. ABI compatibility is offered in languages which require it.
+
+**SDK Stability -**
+Public portions of the SDK (constructors, configuration, end-user interfaces) must remain backwards compatible. Internal interfaces are allowed to break; ABI compatibility is not required.
+
+**Semantic Conventions Stability -**
+Semantic Conventions may not be removed once they are stable. New conventions may be added to replace usage of older conventions, but the older conventions are never removed, they will only be marked as deprecated in favor of the newer ones.
+
+**Contrib Stability -**
+Plugins and instrumentation are kept up to date, and are released simultaneously (or shortly after) the latest release of the API. The goal is to ensure users can update to the latest version of opentelemetry, and not be held back by the plugins that they depend on.
+
+Public portions of contrib packages (constructors, configuration, interfaces) must remain backwards compatible. ABI compatibility not required.
+
+Telemetry produced by contrib instrumentation must also remain stable and backwards compatible, to avoid breaking alerts and dashboard.
+
+### Deprecation
+
+In theory, signals could be replaced. When this happens, they are marked as deprecated.
+
+Code is only marked as deprecated when the replacement becomes stable. Deprecated code still abides by the same support guarantees as stable code. Deprecated APIs remain stable and backwards compatible.
+
+### Removal
+
+Packages are end-of-life’d by being removed from the release. The release then makes a major version bump.
+
+We currently have no plans for deprecating signals or creating a major version past v1.0.
+
+## Version Numbers
+
+OpenTelemetry follows [semver 2.0](https://semver.org/) conventions, with the following distinction.
+
+For the purposes of versioning, all code within a release is treated as if it were part of a single package. All packages within a release receive the same version number. By keeping version numbers in lockstep, end users can understand which packages belong to which release.
+
+Note that contrib packages are released separately from core packages. Contrib releases are not required to match version numbers with the core releases.
+
+In some languages, package managers react poorly to experimental packages having a version higher than 0.X. In these cases, a language-specific workaround is required.
```

Comment:
What are these languages? SemVer 2.0 describes that unstable / pre-release / experimental package versions could be expressed as `vX.Y.Z` then `-rc0` or `-build123`. Even the tools that are not fully SemVer 2.0 compliant, e.g. NuGet package manager - generally respects the experimental packages, where you can label these as `v1.2.3-build333` ... Would you be able to share concrete examples? Is there a plan in these languages to start supporting SemVer 2.0?

Reply:
Go, specifically. Go maintainers are concerned that experimental packages with versions higher than v0.X will be mishandled by the go module system. 

I think it would be great to have one release version, but use -alpha or similar delimiter for the experimental packages. That would add clarity to users, but I can also see how some package managers could be confused and think those packages are part of a separate release. but if it works, I think it would be helpful to do this.

I would love some clarity on this subject, as I am not experienced with every package manager currently in use by otel.

Reply:
It'd be a bit unnatural for C++ and C# packages to have `-experimental` in the package name. It is also becoming a huge hassle for a project to change the package name in their configuration from `opentelemetry-sdk` to `opentelemetry-sdk-experimental`. Instead of going that route, of having separate package names - I'd rather suggest that we allow for flexibility here.

For example, .NET recommendations for `experimental` or `alpha` package versions from here:
https://docs.microsoft.com/en-us/nuget/create-packages/prerelease-packages

![image](https://user-images.githubusercontent.com/34072974/101839780-d7653780-3af7-11eb-9b9c-b174a41eb2cf.png)

Specifically including the `-alpha` or `-experimental` or `-rc1` as part of **Package Version** rather than as part of package name. This versioning strategy also makes sense for C++ packages and Linux packages. It's also part of SemVer 2.0 guidelines. Splitting from 2 package sets (`sdk`, `api`) into 4 package sets (`sdk`,`sdk-experimental`,`api`,`api-experimental`) is generally NOT what is a common practice for C# and C++ packages, nor what SemVer 2.0 is suggesting.

Can we have provision for some flexibility in this regard? `Go` can do what is within their constraints, whereas for C# and for C++ (and I believe for Java too), it'd make practical sense to include the alpha or pre-release designation as part of Package Version, maintaining the same package name.

From a practical standpoint, I'd see how my customers in Azure cloud are gonna scream loud if I ask them to change the `Package Name` to add `-experimental`, and they'd generally be ok to change the `Package Version` to something that includes `-alpha` or `-rc1` or alike.

Reply:
> Can we have provision for some flexibility in this regard? Go can do what is within their constraints, whereas for C# and for C++

This makes total sense (with the specific case of .Net using `-alpha` in the version rather than in the name), as it complies with what is expected in the .Net world.

In the example, we would want to have a separate `sdk-metrics, version=1.0.0-alpha` package/assembly, more than having `sdk-experimental, version=1.0.0`, for example.

Reply:
I feel for C++ specifically, the term `package` is big vague to describe. There is no centralized packaging system or standard source code package format  (unlike what we have pypi in python, npm for js/nodejs, nuget for C# ... ). We do have third party package managers like vcpkg and canon which we should support on need basis.  Even it is not possible to distribute in binary format, as it is practically not possible to support all platform ( not thinking of ABI for now which is separate discussion).

So the standard distribution would be to release entire source in zip/tar format, which is already done through GitHub release assets. 

We can take advantage of this source distribution, and embed the experimental features/signals behind feature flag ( using standard C `#ifdef` macros ) instead of having a separate package/assembly for these signals. So,(say)  `opentelemetry-sdk v1.0.0` will contain stable version of traces, and experimental logs/metrics behind feature flag. The user then need to explicitly enable these feature flag to use the experimental features.

Reply:
@lalitb - irrespective of what `package` is, I think we can assume that in C++ land we have two dimensions:
- `Package Name`
- `Package Version`

with C++ package name being something like metapackage of `opentelemetry` = `opentelemetry-api` + `opentelemetry-sdk` (whether we actually have to have these split into two packages, good question), and that metapackage would allow a vendor to deliver their own `openelemetry-sdk-${vendor}` --- since we only guarantee ABI stability at API level, and do not guarantee ABI stability for internal SDK interfaces / custom vendor exporters.

My metapoint was that for minor increments and for pre-release packages, we want to adjust the `Package Version` not the `Package Name`. Basically in C++ scenario - `Package Name` is constant, irrelevant whether it is `opentelemetry` as one or `opentelemetry-api` and `opentelemetry-sdk` , plus `opentelemetry-sdk-$vendor`... Basically I want to avoid creating a hassle of maintaining yet another set of `-experimental` packages for us ;-)

Reply:
@maxgolov I completely agree that package **should not move** when they become stable, or otherwise break beta/rc users in an unnecessary fashion. That would be a terrible experience, and would leave users stranded on the experimental package. I've clarified this in the latest update.